### ISSUE TITLE
Stock/Fleet NPCs use more weapons than specified in their ship/variant definition

### ISSUE THREAD

**Issue by tehhowch** [endless-sky/endless-sky#3358]
I've traced this issue to commit https://github.com/endless-sky/endless-sky/commit/ee9d339ffc1a8a52ca2ff37dfbaf32d96f778cda
If the underlying issue persists prior to this commit, it is caught and handled before the ship is able to interact with the player. After this commit though, it is not caught and can leak to the player.

To reproduce:
Fly to planet.
Land.
Close game.
Reopen game.

If any of the mission NPCs specified a variant that does not use all of its hardpoints, you will receive a warning like this one during the "Reopen game." step.
![image](https://user-images.githubusercontent.com/20871346/33535580-fe0c3b48-d873-11e7-92c5-883b0d6fa7df.png)
If you inspect your save without performing any actions with the game, you can observe the faulty definition. Once you trigger a new save generation, the erroneous definition(s) will be stripped.

This is because there are two paths to loading a ship for a mission's NPC:
1) Instantiating a picked ship
2) Loading a completed ship
When your save is loaded, the 2nd path is called for both accepted missions, and available missions.
When you land, the 1st path is called for newly created missions, and no change is done to currently accepted missions.

The weapon that is replicated to all open hardpoints is the outfit that was first globally referenced, e.g when sorted by a `map` container, it compares as less than to the other weapons.

**Comment by tehhowch** [comment]
Here's one of them in my save, before it gets overwritten:
```
"available mission" "Remnant: Defense 1"
	...
	npc
			ship Starling
			name "Sunlit Ocean Remade"
			outfits
				"Epoch Cell"
				"Crystal Capacitor"
				"Quantum Key Stone"
				"Forge-Class Thruster"
				"Forge-Class Steering"
				Hyperdrive
				"Thrasher Cannon" 3
				"Millennium Cell"
				"Thrasher Turret"
			engine -16 82 0.7
			engine 16 82 0.7
			engine 0 90 1
			gun 0 -99 "Thrasher Cannon"
			gun -12 -72 "Thrasher Cannon"
			gun 12 -72 "Thrasher Cannon"
			gun -18 -49 "Thrasher Cannon"
			gun 18 -49 "Thrasher Cannon"
			turret 0 2 "Thrasher Turret"
```
The NPC spec in this mission is:
https://github.com/endless-sky/endless-sky/blob/224e62f563a55fc57e20b2e851a945836378bd0f/data/remnant%20missions.txt#L97-L100
And that fleet spec has several options. Based on the ships selected, this specific instance uses these variants:
https://github.com/endless-sky/endless-sky/blob/224e62f563a55fc57e20b2e851a945836378bd0f/data/remnant.txt#L48-L50
Which are:
https://github.com/endless-sky/endless-sky/blob/224e62f563a55fc57e20b2e851a945836378bd0f/data/remnant%20ships.txt#L118-L136
and
https://github.com/endless-sky/endless-sky/blob/224e62f563a55fc57e20b2e851a945836378bd0f/data/remnant%20ships.txt#L229-L249
and launching the game gave this warning:
![image](https://user-images.githubusercontent.com/20871346/33536041-42a6291e-d877-11e7-9a25-32fe9c466b78.png)

I'll see how far back this goes...

**Comment by tehhowch** [comment]
Looks like this started happening (or at least, started being written to the save file) with ee9d339

**Comment by tehhowch** [comment]
While the ship reports the outfitted armament when scanned, in practice it uses every hardpoint:
scan:
![image](https://user-images.githubusercontent.com/20871346/33539297-57a8f846-d88b-11e7-9d57-6cf8c2ec6a8b.png)


battle:
![image](https://user-images.githubusercontent.com/20871346/33539257-1ccd109a-d88b-11e7-865f-d8c812b29846.png)
This discrepancy persists until the player reloads the game.

**Comment by tehhowch** [comment]
Definitely a tricky bug:
1) Every ship and variant is loaded by the game with the correct `outfits` list, `equipped` list, and `armament` configuration.
2) At the end of `Ship::FinishLoading`, the `equipped` map is cleared (dating back to 28e86cce4e23a9e71324b1569e7b5b76c477a05e )
3) When a mission NPC is instantiated as a 'stockShip' or fleet member, even if it is a variant (e.g. `"Starling (Thrasher)"`), it is treated as a 'base model.'
4) Because this ship is a 'base model', it does not conduct a hardpoint reassignment check, in which the `equipped` map would be rebuilt.
5) Because the equipped map is empty, no 'excess' checks are conducted (which is not the issue, since all armaments, etc were properly loaded when the game started up).
6) The bug: when looping all outfits the mission NPC ship possesses, the game checks to see if any weapons are installed, but were not placed in a hardpoint. This check references the `equipped` map - which is empty for mission NPCs at this point (but not empty during the loading process), and since any weapons that exist cannot be found already equipped, they are re-added, until either they are all used up, or all the hardpoints are full. 

There are at least two ways to fix this:
1) Reexamine why we call Ship::FinishLoading() when instantiating a mission NPC.
2) Avoid clearing the equipped map when exiting FinishLoading()
3) Reference the `armament` count of a given outfit instead of `equipped`

While I think option 2 is the easiest fix, I'm preferring option 1
As near as I can tell, NPC calls Ship::FinishLoading to cripple unrepaired `derelict` ships (because the stock version is full health). If this is truly the only need for Ship::FinishLoading, then a less intrusive function should be used for accomplishing the task of personality-dependent ship status modification.

**Comment by endless-sky** [comment]
The call to FinishLoading() was needed for ships that are defined within the NPC block; it was moved from above in commit ee9d339f.

This issue ought to be fixed by 2633668.

**Comment by tehhowch** [comment]
Can confirm, resolved now.
